plenoptic/simulate/models/frontend.py

```diff
@@ -1,96 +1,407 @@
+
+    Parameters
+    ----------
+    center:
+        Dictates whether center is on or off. The surround always will be the opposite
+        of the center. Must be either ['on', 'off']; default is 'on' center.
+    kernel_size: Union[int, Tuple[int, int]], optional
+    width_ratio_limit:
+        Ratio of surround stdev over center stdev. Surround stdev will be clamped to
+        ratio_limit times center_std.
+    amplitude_ratio:
+        Ratio of center/surround amplitude. Applied before filter normalization.
+    center_std:
+        Standard deviation of circular Gaussian for center.
+    surround_std:
+        Standard deviation of circular Gaussian for surround. Must be at least
+        ratio_limit times center_std.
+    pad_mode:
+        Padding for convolution, defaults to "circular".
+    """
+
+    def __init__(
+        self,
+        kernel_size: Union[int, Tuple[int, int]],
+        center: str = "on",
+        width_ratio_limit: float = 4.0,
+        amplitude_ratio: float = 1.25,
+        center_std: float = 1.0,
+        surround_std: float = 4.0,
+        pad_mode: str = "circular",
+    ):
+        super().__init__()
+
+        assert center in ["on", "off"], "center must be 'on' or 'off'"
+        assert width_ratio_limit > 1.0, "stdev of surround must be greater than center"
+        assert amplitude_ratio >= 1.0, "amplitudes must"
+
+        self.center = center
+        self.kernel_size = kernel_size
+        self.width_ratio_limit = width_ratio_limit
+        self.register_buffer("amplitude_ratio", torch.tensor(amplitude_ratio))
+
+        self.center_std = nn.Parameter(torch.tensor(center_std))
+        self.surround_std = nn.Parameter(torch.tensor(surround_std))
+
+        self.pad_mode = pad_mode
+
+    def get_filter(self) -> Tensor:
+        """Creates an on center/off surround, or off center/on surround conv filter"""
+        filt_center = circular_gaussian(self.kernel_size, self.center_std)
+        filt_surround = circular_gaussian(self.kernel_size, self.surround_std)
+        on_amp = self.amplitude_ratio
+
+        if self.center == "on":  # on center, off surround
+            filt = on_amp * filt_center - filt_surround  # on center, off surround
+        else:  # off center, on surround
+            filt = on_amp * filt_surround - filt_center
+
+        filt = filt / filt.sum()
+
+        return filt.view(1, 1, *filt.shape)
+
+    def _clamp_surround_std(self) -> Tensor:
+        """Clamps surround standard deviation to ratio_limit times center_std"""
+        return self.surround_std.clamp(
+            min=self.width_ratio_limit * float(self.center_std), max=None
+        )
+
+    def forward(self, x: Tensor) -> Tensor:
+        x = same_padding(x, self.kernel_size, pad_mode=self.pad_mode)
+        self._clamp_surround_std()  # clip the surround stdev
+        filt = self.get_filter()
+        y = F.conv2d(x, filt, bias=None)
+        return y
+
+
+class LN(nn.Module):
```

Comment:
We have both because this entire file is based on Alex's models and includes all models he used in his paper.

One main difference is that LinearNonlinear has no nullspace has>1 output channel, whereas frontend.LN does have a null space and only one channel output.

I do agree it is confusing.